File: src/main/java/edu/eci/arsw/blacklistvalidator/Supervisor.java
package edu.eci.arsw.blacklistvalidator;

import edu.eci.arsw.spamkeywordsdatasource.HostBlacklistsDataSourceFacade;

import static edu.eci.arsw.blacklistvalidator.HostBlackListsValidator.*;

/*
 *
 * @author Daniel palacios Moreno & Sofia Nicolle Ariza Goenaga
 */
public class Supervisor extends Thread{
    private String host;
    private int section;
    private int sectionSize;

    public Supervisor(String host, int section, int sectionSize){
        this.host = host;
        this.section = section;
        this.sectionSize = sectionSize;
    }

    public void run(){
        try {
            HostBlacklistsDataSourceFacade skds = HostBlacklistsDataSourceFacade.getInstance();

            for (int i = section * sectionSize; i < (section + 1) * sectionSize; i++){
                if (!shouldContinueSearching()) {
                    break;
                }

                incrementCheckedCount();
                if (skds.isInBlackListServer(i, host)){
                    addBlackListOcurrence(i);
                }
            }
        } finally {
            notifyThreadCompletion();
        }
    }
}

File: goImplementation/supervisor.go
package main

// Supervisor represents a goroutine that checks a section of blacklist servers
// Author: Daniel palacios Moreno & Sofia Nicolle Ariza Goenaga
type Supervisor struct {
	host        string
	section     int
	sectionSize int
	validator   *HostBlackListsValidator
}

// NewSupervisor creates a new Supervisor instance
func NewSupervisor(host string, section int, sectionSize int, validator *HostBlackListsValidator) *Supervisor {
	return &Supervisor{
		host:        host,
		section:     section,
		sectionSize: sectionSize,
		validator:   validator,
	}
}

// Run executes the supervisor's task of checking blacklist servers
func (s *Supervisor) Run() {
	defer s.validator.NotifyThreadCompletion()

	skds := GetHostBlacklistsDataSourceFacadeInstance()

	start := s.section * s.sectionSize
	end := (s.section + 1) * s.sectionSize

	for i := start; i < end; i++ {
		if !s.validator.ShouldContinueSearching() {
			break
		}

		s.validator.IncrementCheckedCount()
		if skds.IsInBlackListServer(i, s.host) {
			s.validator.AddBlackListOccurrence(i)
		}
	}
}
